numpy/core/defchararray.py

```diff
@@ -2092,6 +2093,14 @@ def __new__(subtype, shape, itemsize=1, unicode=False, buffer=None,
         _globalvar = 0
         return self
 
+    def __array_prepare__(self, arr, context=None):
+        if arr.dtype.char in "US":
+            return arr.view(type(self))
+        return arr
+
+    def __array_wrap__(self, arr, context=None):
+        return arr
```

Comment:
If we need this, we can probably do without array-prepare.  And we should use the logic from there here.  A couple of functions will use array-wrap to return the subclass, and you are disabling that by adding this (I think).  For example:
```
In [1]: class myarr(np.ndarray):
   ...:     pass
   ...: 

In [2]: a = np.arange(3).view(myarr)

In [3]: 

In [3]: np.delete(a, 0)
Out[3]: myarr([1, 2])

In [4]: class myarr(np.ndarray):
   ...:     def __array_wrap__(self, arr, context=None):
   ...:         return arr
   ...: 

In [5]: a = np.arange(3).view(myarr)

In [6]: np.delete(a, 0)
Out[6]: array([1, 2])
```
So if array-wrap works better (kicks in earlier/also), then that is fine.  But it should at least retain the type when possible (which is its job!).

Reply:
Removing array-prepare fails again though.

```diff
numpy on  string-ufuncs-isalpha [$!?] is 📦 v2.0.0.dev0 via 🐍 pyenv mambaforge (venv) took 8s 
❯ git diff             
diff --git a/numpy/core/defchararray.py b/numpy/core/defchararray.py
index b5086b4af..9ed7f471a 100644
--- a/numpy/core/defchararray.py
+++ b/numpy/core/defchararray.py
@@ -2093,15 +2093,14 @@ def __new__(subtype, shape, itemsize=1, unicode=False, buffer=None,
         _globalvar = 0
         return self
 
-    def __array_prepare__(self, arr, context=None):
+    def __array_wrap__(self, arr, context=None):
+        print(f"__array_wrap__ called with {self=}, {arr=}, {arr.dtype.char=}")
         if arr.dtype.char in "US":
             return arr.view(type(self))
         return arr
 
-    def __array_wrap__(self, arr, context=None):
-        return arr
-
     def __array_finalize__(self, obj):
+        print(f"__array_finalize__ called with {self=}, {obj=}, {obj.dtype.char=}")
         # The b is a special case because it is used for reconstructing.
         if not _globalvar and self.dtype.char not in 'SUbc':
             raise ValueError("Can only create a chararray from string data.")
```
```python3
numpy on  string-ufuncs-isalpha [$!?] is 📦 v2.0.0.dev0 via 🐍 pyenv mambaforge (venv) took 5s 
❯ python -m spin python
Invoking `build` prior to invoking Python:
/Users/lysnikolaou/repos/python/numpy/venv/lib/python3.11/site-packages/spin/cmds/meson.py:99: ResourceWarning: unclosed file <_io.TextIOWrapper name='build/meson-info/meson-info.json' mode='r' encoding='UTF-8'>
  meson_info = json.load(open(meson_info_fn))
ResourceWarning: Enable tracemalloc to get the object allocation traceback
$ /Users/lysnikolaou/repos/python/numpy/venv/bin/python vendored-meson/meson/meson.py compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /Users/lysnikolaou/repos/python/numpy/venv/bin/ninja -C /Users/lysnikolaou/repos/python/numpy/build
ninja: Entering directory `/Users/lysnikolaou/repos/python/numpy/build'
[1/1] Generating numpy/generate-version with a custom command
Saving version to numpy/version.py
$ /Users/lysnikolaou/repos/python/numpy/venv/bin/python vendored-meson/meson/meson.py install --only-changed -C build --destdir ../build-install
$ export PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
🐍 Launching Python with PYTHONPATH="/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages"
$ /usr/bin/env python -P
Python 3.11.3 (tags/v3.11.3:f3909b8bc8, May 29 2023, 12:51:44) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> a = np.char.array(['abc', 'def'])
__array_finalize__ called with self=chararray(['abc', 'def'], dtype='<U3'), obj=array(['abc', 'def'], dtype='<U3'), obj.dtype.char='U'
__array_finalize__ called with self=chararray(['', ''], dtype='<U3'), obj=chararray(['abc', 'def'], dtype='<U3'), obj.dtype.char='U'
>>> a.isalpha()
__array_finalize__ called with self=chararray([False,  True]), obj=chararray(['abc', 'def'], dtype='<U3'), obj.dtype.char='U'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages/numpy/core/defchararray.py", line 2394, in isalpha
    return isalpha(self)
           ^^^^^^^^^^^^^
  File "/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages/numpy/core/defchararray.py", line 866, in isalpha
    return multiarray.isalpha(a)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/lysnikolaou/repos/python/numpy/build-install/usr/lib/python3.11/site-packages/numpy/core/defchararray.py", line 2106, in __array_finalize__
    raise ValueError("Can only create a chararray from string data.")
ValueError: Can only create a chararray from string data.
```

I though that since we choose whether to return the subclass in array-prepare, we'd be fine with just removing the passed in array in array-wrap. Is that not the case?

Reply:
So you are saying both array-prepare and array-wrap get used and at least for ufuncs the special array-prepare is required?  That is fine, I just didn't (and still don't) understand why the `__array_wrap__` is needed for ufuncs, and it looked wrong.  It is also fine to just use the same function for both, i think.

Reply:
While I don't understand it either, it seems like both are needed, yes. I've adjusted the code so that both do the exact same thing.